File: 240929/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.py
from collections import deque

n, m = map(int, input().split())
area = []
for i in range(n):
    area.append(list(map(int, input().split())))
visited = []
for i in range(n):
    visited.append([0] * m)

def in_range(x, y):
    if 0 <= x < n and 0 <= y < m:
        return True
    return False

def can_go(x, y):
    if in_range(x, y) and area[x][y] == 1 and not visited[x][y]:
        return True
    return False

q = deque()
q.append([0, 0])

ans = 0
def bfs():
    global ans
    while q:
        tmp = q.popleft()
        x, y = tmp[0], tmp[1]

        if x == n - 1 and y == m - 1:
            ans = 1

        dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]

        for dr in range(4):
            nx, ny = x + dx[dr], y + dy[dr]

            if can_go(nx, ny):
                visited[nx][ny] = 1
                q.append([nx, ny])

visited[0][0] = 1
bfs()
print(ans)

File: 250223/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.cpp
#include <iostream>
#include <queue>
using namespace std;

int n, m;
int area[100][100];
queue<pair<int, int>> q;
bool visited[100][100];
int ans = 0;

int dx[] = { -1, 0, 1, 0 };
int dy[] = { 0, 1, 0, -1 };

bool InRange(int x, int y) {
    return 0 <= x && x < n && 0 <= y && y < m;
}

bool CanGo(int x, int y) {
    return InRange(x, y) && area[x][y] == 1 && !visited[x][y];
}

void BFS() {
    while (!q.empty()) {
        pair<int, int> v = q.front();
        int x = v.first, y = v.second;
        q.pop();

        for (int dr = 0; dr < 4; dr++) {
            int nx = x + dx[dr];
            int ny = y + dy[dr];

            if (nx == n - 1 && ny == m - 1) {
                ans = 1;
                return;
            }

            if (CanGo(nx, ny)) {
                visited[nx][ny] = true;
                q.push({ nx, ny });
            }
        }
    }
}

int main() {
    cin >> n >> m;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            cin >> area[i][j];

    q.push({ 0, 0 });
    BFS();

    cout << ans;
    return 0;
}